### PR TITLE
8252476: as_Worker_thread() doesn't check what it intends

### PR DESCRIPTION
**Pull request by albertnetymk**
Convert `as_Worker_thread` to non-virtual, like other `as_*_thread` methods.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252476](https://bugs.openjdk.java.net/browse/JDK-8252476): as_Worker_thread() doesn't check what it intends

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Contributors
 * David Holmes `<dholmes@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4167/head:pull/4167` \
`$ git checkout pull/4167`

Update a local copy of the PR: \
`$ git checkout pull/4167` \
`$ git pull https://git.openjdk.java.net/jdk pull/4167/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4167`

View PR using the GUI difftool: \
`$ git pr show -t 4167`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4167.diff">https://git.openjdk.java.net/jdk/pull/4167.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -1589,6 +1589,11 @@ inline JavaThread* JavaThread::current() {
   return Thread::current()->as_Java_thread();
 }
 
+inline WorkerThread* Thread::as_Worker_thread() const {
+  assert(is_Worker_thread(), "Dubious cast to WorkerThread*?");
+  return (WorkerThread*)this;
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:1594]: Can you use static_cast here rather the C-style cast please.

File: src/hotspot/share/runtime/thread.hpp [hunk 2]
```diff
@@ -1589,6 +1589,11 @@ inline JavaThread* JavaThread::current() {
   return Thread::current()->as_Java_thread();
 }
 
+inline WorkerThread* Thread::as_Worker_thread() const {
+  assert(is_Worker_thread(), "Dubious cast to WorkerThread*?");
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:1593]: More than dubious I would say, just plain wrong.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Albert,

Thanks for fixing this. A couple of nits on the details.

David

File: src/hotspot/share/runtime/thread.hpp (hunk 1 above)
Comment (albertnetymk) [reply src/hotspot/share/runtime/thread.hpp:1594]: I get an compiling error, `static_cast from 'const Thread *' to 'WorkerThread *', which are not related by inheritance, is not allowed`. As `WorkerThread` is only declared in this context, the compiler doesn't know the inheritance relation here.

File: src/hotspot/share/runtime/thread.hpp (hunk 2 above)
Comment (albertnetymk) [reply src/hotspot/share/runtime/thread.hpp]: Fixed.

**Review by dholmes-ora — APPROVED** [review]

**Comment by albertnetymk** [comment]
Moved the definition to the inline file so that `static_cast` can be used.

**Review by dholmes-ora — COMMENTED** [review]
The update seems fine to me. But you need to check with @iklam  whether including nonJavaThread.hpp in thread.inline.hpp negates the point of splitting out nonJavaThread.hpp in the first place.

Thanks,
David

**Comment by iklam** [comment]
> The update seems fine to me. But you need to check with @iklam whether including nonJavaThread.hpp in thread.inline.hpp negates the point of splitting out nonJavaThread.hpp in the first place.
> 

I think it's OK to include nonJavaThread.hpp in thread.inline.hpp (included by about 30% of hotspot .o files). But we should avoid including it in thread.hpp (included by about 80% of hotspot .o files)

BTW, we should make `JavaThread::as_CompilerThread` consistent as well: move to thread.inline.hpp, change the casting,  and rename to `as_Compiler_thread` (in another RFE??)

**Review by iklam — APPROVED** [review]

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by albertnetymk** [comment]
> BTW, we should make JavaThread::as_CompilerThread consistent as well: move to thread.inline.hpp, change the casting, and rename to as_Compiler_thread (in another RFE??)

I can do so in another PR. Thank you for the suggestion.

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**